sa/sa.go

```diff
@@ -1812,17 +1812,43 @@ func (ssa *SQLStorageAuthority) GetOrderForNames(
 	// Hash the names requested for lookup in the orderFqdnSets table
 	fqdnHash := hashNames(req.Names)
 
-	var orderID int64
-	err := ssa.dbMap.WithContext(ctx).SelectOne(&orderID, `
-	SELECT orderID
-	FROM orderFqdnSets
-	WHERE setHash = ?
-	AND registrationID = ?
-	AND expires > ?`,
-		fqdnHash, *req.AcctID, ssa.clk.Now())
-
-	// There isn't an unexpired order for the provided AcctID that has the
-	// fqdnHash requested.
+	// Find a possibly-suitable order. We don't include the account ID or order
+	// status in this query because there's no index that includes those, so
+	// including them could require the DB to scan extra rows.
+	// Instead, we select one unexpired order that matches the fqdnSet. If
+	// that order doesn't match the account ID or status we need, just return
+	// nothing. We use `ORDER BY expires ASC` because the index on
+	// (setHash, expires) is in ASC order. DESC would be slightly nicer from a
+	// user experience perspective but would be slow when there are many entries
+	// to sort.
+	// This approach works fine because in most cases there's only one account
+	// issuing for a given name. If there are other accounts issuing for the same
+	// name, it just means order reuse happens less often.
+	var result struct {
+		OrderID        int64
+		RegistrationID int64
+	}
+	var err error
+	if features.Enabled(features.FasterGetOrderForNames) {
```

Comment:
It looks like this diff lost the part in https://github.com/letsencrypt/boulder/pull/4326 where the `RegistrationID` is enforced as correct when `features.FasterGetOrderForNames` is enabled.

In this branch the new query doesn't have a `WHERE registrationID = ?` clause, but it also doesn't check that the `result.RegistrationID` is equal to `*req.AcctID`.

Reply:
Good catch, fixed!